sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/postgreSQL/PostgreCastBase.scala

```diff
@@ -68,16 +91,47 @@ case class PostgreCastToBoolean(child: Expression, timeZoneId: Option[String])
             throw new IllegalArgumentException("invalid input syntax for type boolean: $c");
           }
         """
-
     case IntegerType =>
       super.castToBooleanCode(from)
   }
+}
 
-  override def dataType: DataType = BooleanType
+case class PostgreCastToTimestamp(child: Expression, timeZoneId: Option[String])
+  extends PostgreCastBase(TimestampType) {
 
-  override def nullable: Boolean = child.nullable
+  override def fromTypes: TypeCollection = TypeCollection(StringType, DateType, NullType)
 
-  override def toString: String = s"PostgreCastToBoolean($child as ${dataType.simpleString})"
+  override def withTimeZone(timeZoneId: String): TimeZoneAwareExpression =
+    copy(timeZoneId = Option(timeZoneId))
 
-  override def sql: String = s"CAST(${child.sql} AS ${dataType.sql})"
+  override def castToTimestamp(from: DataType): Any => Any = from match {
+    case StringType =>
+      buildCast[UTF8String](_, utfs => DateTimeUtils.stringToTimestamp(utfs, zoneId)
```

Comment:
I believe that postgre could correctly parse string `19700101`, `1970/01/01`, `January 1 04:05:06 1970 PST` while spark can't. So, I think that we may also need to support it in `PostgreCastToTimestamp`.

Reply:
Thanks for your suggestion.  I will check this.

Reply:
```
postgres# select cast('19700101' as timestamp);
01.01.1970 00:00:00
```
```
postgres# select cast('1970/01/01' as timestamp);
01.01.1970 00:00:00
```
```
postgres# select cast('January 1 04:05:06 1970 PST' as timestamp);
01.01.1970 04:05:06
```
Spark results with NULL for all of them.

Reply:
@maropu kindly review latest changes and give your feedback on supporting above queries.

Do we need to support them in this PR? If yes, we need to list all formats for timestamps which postgres supports but spark don't.

Reply:
I personally think that the support above is not a main issue of this pr, so better to separate the two work: the timestamp cast support and the timestamp format support for the pg dialect.